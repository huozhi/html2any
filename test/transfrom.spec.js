const React = require('react')
const renderer = require('react-test-renderer')
const parse = require('../src/parse')
const transform = require('../src/transform')
const {html1, html2} = require('./contents')

function rule(node, children) {
  if (typeof node === 'string') {
    return node
  }
  const {name, attributes} = node
  if (attributes && attributes.hasOwnProperty('style')) {
    delete attributes.style
  }
  if (attributes && attributes.hasOwnProperty('class')) {
    attributes.className = attributes.class
    delete attributes.class
  }
  let elem
  switch (name) {
    case 'p': {
      elem = {
        type: 'h1',
        props: attributes,
      }
      break
    }
    case 'b': {
      elem = {
        type: 'h2',
        props: attributes,
      }
      break
    }
    case 'div': {
      elem = {
        type: 'div',
        props: attributes,
      }
      break
    }
    case 'input': {
      if (attributes.readonly) {
        attributes.readOnly = attributes.readonly
        delete attributes.readonly
      }
      elem = {
        type: 'input',
        props: attributes,
      }
    }
    default: {
      elem = {
        type: name,
        props: attributes,
      }
    }
  }

  if (!elem || !elem.type) {
    return null
  }

  // add react index prop
  if (typeof node.index === 'number') {
    Object.assign(elem.props, {key: node.index})
  }
  return React.createElement(elem.type, elem.props, children)
}

it('transform works well on html1 with customized rule', () => {
  const ast = parse(html1)[0]
  const result = transform(ast, rule)
  const output = renderer.create(result).toJSON()
  expect(output).toMatchSnapshot()
})

it('transform works well on html2 with customized rule', () => {
  const ast = parse(html2)[0]
  const result = transform(ast, rule)
  const output = renderer.create(result).toJSON()
  expect(output).toMatchSnapshot()
})
