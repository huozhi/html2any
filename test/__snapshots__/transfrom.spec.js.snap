// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform works well on html1 with customized rule 1`] = `
<div>
  <hr />
  <img
    data-index="2"
    src="a.jpg"
  />
  <h1
    id="x34"
  >
    1234 &amp;
  </h1>
  <div
    className="d3 d2"
    id="container"
  >
    <input
      data-id="xx"
      readOnly={true}
      value="1 2 3"
    />
  </div>
</div>
`;

exports[`transform works well on html2 with customized rule 1`] = `
<div>
  <h1>
    <br />
  </h1>
  <h1
    data-mce-style="text-align: center; font-size: 15px;"
  >
    <img
      alt="Tiny M C E"
      data-mce-src="images/glyph-tinymce@2x.png"
      height="97"
      src="https://www.tinymce.com/images/glyph-tinymce@2x.png"
      width="110"
    />
  </h1>
  <h1>
    This demo includes these TinyMCE Premium
    <span
      className="mce-spellchecker-word"
      data-mce-bogus="1"
      data-mce-index="0"
      data-mce-word="Plugins"
    >
      Plugins
    </span>
    :
    <a
      data-mce-href="powerpaste"
      data-mce-linkchecker-status="valid"
      href="https://www.tinymce.com/powerpaste"
    >
      PowerPaste
    </a>
    (improved Word paste),
    <a
      data-mce-href="accessibility-checker"
      data-mce-linkchecker-status="valid"
      href="https://www.tinymce.com/accessibility-checker"
    >
      Accessibility Checker
    </a>
    ,
    <a
      data-mce-href="spell-checker-pro"
      data-mce-linkchecker-status="valid"
      href="https://www.tinymce.com/spell-checker-pro"
    >
      Spell Checker Pro
    </a>
    &amp; file management with
    <a
      data-mce-href="http://www.moxiemanager.com"
      data-mce-linkchecker-status="valid"
      href="http://www.moxiemanager.com"
    >
      MoxieManager
    </a>
    .
  </h1>
  <h1
    data-mce-style="text-align: center; color: #666; font-size: 15px; font-family: avenir; font-weight: 200;"
  >
    TinyMCE is a platform indepen
    <strong>
      dent web-
    </strong>
    based JavaScript HTML WYSIWYG
    <br />
    editor control released as open source
    <em>
      under
    </em>
    LGPL.
  </h1>
  <h1
    data-mce-style="text-align: center; color: #666; font-size: 15px; font-family: avenir; font-weight: 200;"
  >
    <em>
      TinyMCE enables you to convert HTML
      <span
        className="mce-spellchecker-word"
        data-mce-bogus="1"
        data-mce-index="0"
        data-mce-word="textarea"
      >
        textarea
      </span>
      fields or other HTML elements to editor instances.
    </em>
  </h1>
  <h1>
    <br />
  </h1>
</div>
`;
